MAME Fighting Game Controls

TVsIan opened the pull request:
Alternate control schemes for 5/6 button arcade games.

* Adds three alternate control schemes based on the SNES controller (to match the standard Retropad layout):
* Street Fighter II style - "top" face buttons are punches, "bottom" are kicks, strong attacks on L/R.
* Mortal Kombat style - "left" face buttons are punches, "right" face buttons are kicks, block/run on L/R.
* Killer Instinct style - same as Street Fighter II, but with the fast/weak attacks on L/R.
* Auto behavior (option not set) will compare against a list of ROMs and apply the matching scheme.
* Can be manually overridden at game or system level, including the default Batocera MAME controls.
* Will not conflict with controls that are manually set, it changes the default while generating the config, but per-game controls will override.

nadenislamarre commented:
hum. eventulaly, can we do an external .txt file ? to not polluate to much the configgen code ?

TVsIan commented:
It might make sense to spin the controller config off into a separate
script, I know some of the other generators do it that way. It's not
something I really have experience doing, but I could probably figure it
out if needed.

On Fri, Dec 3, 2021 at 12:48 PM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> hum. eventulaly, can we do an external .txt file ? to not polluate to much
> the configgen code ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-985712032>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2U4INJJGLX5R4VLDWDUPD7GNANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

Hew-ux commented:
Awesome work! This saves so much time configuring the controls for fighting games.

TVsIan commented:
> Awesome work! This saves so much time configuring the controls for fighting games.
> 
> While you're at it, do you think there's a way you could implement this as well: [controls.zip](https://github.com/batocera-linux/batocera.linux/files/7653829/controls.zip) It's a database that gives names to all of the buttons, so they don't appear as "button 1, 2" and instead appear as "light punch, medium punch" etc.
> 
> I've been looking at it myself but have no clue where to start.

As far as I know, controls.xml hasn't been updated in a very long time, and doesn't work inside MAME. It's mostly used for things like LEDBlinky and other frontends. If we were setting controls manually in ES it might make sense to use, and it might be something to use if dynamically generating the controller layout images, but I don't think it would do anything in MAME.

Hew-ux commented:
Ah I see. it's something to do on my "eventually" last once I've worked out automatic controller overlay generation.

lbrpdx commented:
I like this option. What is the status? Still trying to put the configuration in an external file? We could loop controller tattoos into this too, what do you think @Hew-ux ?

Hew-ux commented:
Hm, you mean like somehow get this to specify which controller overlay to use? That sounds cool, but first I need to sort out everything else. it's on the "todo" list.

TVsIan commented:
> I like this option. What is the status? Still trying to put the configuration in an external file? We could loop controller tattoos into this too, what do you think @Hew-ux ?

Currently in the process of splitting the config generation off into a separate script, I may have that done this week.

Hew-ux commented:
One suggestion: Could you implement the "modern fighter" layout that's [built into lr-mame2003plus](https://docs.libretro.com/library/mame2003_plus/#default-retropad-layouts) as one of the options? And then the es_features labels could specify like so:
```
                "MAME default":      0
                "Street Fighter (SNES)":    1
                "Mortal Kombat":     2
                "Killer Instinct":   3
                "Street Fighter (Modern)" : 4
```
In the config, I think that would be:
```
        if altButtons == 4: # Capcom 6-button Mapping (the "modern fightstick" layout used in SFIV and above)
            mappings.update({"BUTTON1": "y"})
            mappings.update({"BUTTON2": "x"})
            mappings.update({"BUTTON3": "pagedown"})
            mappings.update({"BUTTON4": "b"})
            mappings.update({"BUTTON5": "a"})
            mappings.update({"BUTTON6": "r2"})
```
but I'm not sure, please double check that if you're implementing it.

TVsIan commented:
Sure, once I get the script split worked out, can add that in.

On Mon, Dec 13, 2021 at 8:16 PM Atari ***@***.***> wrote:

> One suggestion: Could you implement the "modern fighter" layout that's built
> into lr-mame2003plus
> <https://docs.libretro.com/library/mame2003_plus/#default-retropad-layouts>
> as one of the options? And then the es_features labels could specify like
> so:
>
>                 "MAME default":      0
>                 "Street Fighter (SNES)":    1
>                 "Mortal Kombat":     2
>                 "Killer Instinct":   3
>                 "Street Fighter (Modern)" : 4
>
> In the config, I think that would be:
>
>         if altButtons == 4: # Capcom 6-button Mapping (the "modern fightstick" layout used in SFIV and above)
>             mappings.update({"BUTTON1": "y"})
>             mappings.update({"BUTTON2": "x"})
>             mappings.update({"BUTTON3": "pagedown"})
>             mappings.update({"BUTTON4": "b"})
>             mappings.update({"BUTTON5": "a"})
>             mappings.update({"BUTTON6": "r2"})
>
> but I'm not sure, please double check that if you're implementing it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-993064140>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2RPW6ZQWY7TBDRE4OTUQ2LHDANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

Hew-ux commented:
Time to push boundaries, one for Genesis-style controllers too:
```
                "MAME default":             0
                "Street Fighter (SNES)":    1
                "Mortal Kombat":            2
                "Killer Instinct":          3
                "Street Fighter (Modern)": 4
                "Genesis (6 face buttons)": 5
```
Configgen:
```
        if altButtons == 5: # Genesis-style controller layout
            mappings.update({"BUTTON1": "pageup"})
            mappings.update({"BUTTON2": "x"})
            mappings.update({"BUTTON3": "pagedown"})
            mappings.update({"BUTTON4": "y"})
            mappings.update({"BUTTON5": "x"})
            mappings.update({"BUTTON6": "a"})
```
Totally understandable if this is "too far", though. It would just be nice to be able to use a single Genesis/Megadrive controller for both Street Fighter 2 in arcade and on the Genesis/Megadrive itself.

TVsIan commented:
That would be easy enough too. And handy whenever Retro Bit releases that
Saturn “pro controller” they’ve posted pictures of.

On Mon, Dec 13, 2021 at 8:22 PM Atari ***@***.***> wrote:

> Time to push boundaries, one for Genesis-style controllers too:
>
>                 "MAME default":             0
>                 "Street Fighter (SNES)":    1
>                 "Mortal Kombat":            2
>                 "Killer Instinct":          3
>                 "Street Fighter (Modern)": 4
>                 "Genesis (6 face buttons)": 5
>
> Configgen:
>
>         if altButtons == 5: # Genesis-style controller layout
>             mappings.update({"BUTTON1": "pageup"})
>             mappings.update({"BUTTON2": "x"})
>             mappings.update({"BUTTON3": "pagedown"})
>             mappings.update({"BUTTON4": "y"})
>             mappings.update({"BUTTON5": "x"})
>             mappings.update({"BUTTON6": "a"})
>
> Totally understandable if this is "too far", though. It would just be nice
> to be able to use a single Genesis/Megadrive controller for both Street
> Fighter 2 in arcade and on the Genesis/Megadrive itself.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-993067185>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2UGUL4PSMIFLYLMWV3UQ2L6NANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

TVsIan commented:
I have a functioning version now, with the extra control schemes. I changed the text a little - the existing ones have had (SNES) added to their name, the Street Fighter (Modern) is the same, and the Genesis one is named Genesis 6-Button (Retroarch) to indicate it uses the same mapping as the Retroarch Genesis cores.

I did confirm an issue I had noticed earlier. The MAME control generator can't handle L2/R2 if they're an axis. With an XBox controller, buttons 7 & 8 are "Unknown" in the config. With a fightstick and digital buttons, they map properly. I suspect that it hasn't really come up since most games that people would actually play use 6 buttons or less.

I don't know enough about what pad.input is spitting out for L2/R2. Mapping those controls inside MAME produces an axis (2 & 5) as expected, but the controller name (X360Controller) is actually part of the definition. On the windows side, setting ZAxis + & - in the config seems to set it properly once MAME parses the controller file.

My ordinary debugging method of printing a value to the console won't work for this since it can't print an input type - is there some way to figure out what it's actually trying to write for those controls so I can make a workaround?

Hew-ux commented:
Could you try `evtest` in SSH to see the inputs maybe?

I remember duckstation having a similar problem too.

TVsIan commented:
I'll give that a shot. I just looked up the Duckstation generator, and it looks like it's got some stuff about checking the type of control, one of which is Axis. I'll dig into how the input type is constructed and I may be able to get a workaround in place from that. I suspect that particular part of the generator is looking for buttons and writing a value MAME can't figure out.

TVsIan commented:
Ok, fixed that, but discovered something even stranger. Apparently the generator doesn't get input IDs for joystick right or down, only left and up, presumably due to how the controllers are configured in ES. Looks like the generator has been filling those in. So that's causing a bit of a headache in modifying the MESS control code to have fixed trigger buttons, but I think I can get that fixed (at least so it's functional if not 100% proper) too.

TVsIan commented:
Ok, everything should be fully fixed and ready to go. The input2definition function may need an additional couple of lines for Dualshock/DInput controllers, if those triggers don't work I would just need to know the input.id for the buttons and what MAME sets them to if you map them manually.

ElMeroMaquinero commented:
You are missing Neo-Geo button layout

TVsIan commented:
Well, what would the preferred layout be, and would it apply to all Neo Geo
games? I didn’t realize the default was no good.

On Thu, Dec 16, 2021 at 12:44 PM ElMeroMaquinero ***@***.***>
wrote:

> You are missing Neo-Geo button layout
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996039368>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2SGXJAV3RZIVZJSRMLURIQPJANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.
>

ElMeroMaquinero commented:
See here:  https://images.app.goo.gl/ED9NeUvDPsqc42mM7

TVsIan commented:
Yes, I know what the control panel looks like, but how would that be set to
a SNES controller differently than it does now?

On Thu, Dec 16, 2021 at 12:50 PM ElMeroMaquinero ***@***.***>
wrote:

> See here: https://images.app.goo.gl/ED9NeUvDPsqc42mM7
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996043563>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2Q76RWG2AL2KNTVXD3URIRFBANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ElMeroMaquinero commented:
In an arcade application the button layout standard is 6 button cluster like street fighter.  When playing a Neo Geo game, In my opinion, the button order would feel right as... b, y, x, L,a,R

ElMeroMaquinero commented:
Btw, I was referred to this PR due to the fact that you can’t change the button mappings on the MAME stand alone emulator.  I still feel that this is not a solution.  We do have the ability to custom map on libretro versions.

TVsIan commented:
You can, just not system-wide. Changing the options in the MAME menu under
“this machine” will save them for that game.

But I’ll see what FBNeo does, and maybe set up a default based on that if
it’s different plus one for fightsticks.

On Thu, Dec 16, 2021 at 1:05 PM ElMeroMaquinero ***@***.***>
wrote:

> Btw, I was referred to this PR due to the fact that you can’t change the
> button mappings on the MAME stand alone emulator. I still feel that this is
> not a solution. We do have the ability to custom map on libretro versions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996054673>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2SMGOQTCOFX5FKT7ITURIS6LANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TVsIan commented:
Ok, it looks like there are three control schemes that would make sense to
add. The Neo Geo Mini pad, the Neo Geo CD pad, and fight stick. I’ll add
those and see which default lines up with the controls that ports would
have used.

On Thu, Dec 16, 2021 at 1:20 PM Ian Murray ***@***.***> wrote:

> You can, just not system-wide. Changing the options in the MAME menu under
> “this machine” will save them for that game.
>
> But I’ll see what FBNeo does, and maybe set up a default based on that if
> it’s different plus one for fightsticks.
>
> On Thu, Dec 16, 2021 at 1:05 PM ElMeroMaquinero ***@***.***>
> wrote:
>
>> Btw, I was referred to this PR due to the fact that you can’t change the
>> button mappings on the MAME stand alone emulator. I still feel that this is
>> not a solution. We do have the ability to custom map on libretro versions.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996054673>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AE5KD2SMGOQTCOFX5FKT7ITURIS6LANCNFSM5JKFRBFA>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

ElMeroMaquinero commented:
> You can, just not system-wide. Changing the options in the MAME menu under “this machine” will save them for that game. But I’ll see what FBNeo does, and maybe set up a default based on that if it’s different plus one for fightsticks.
> […](#)

Yes you are correct.  You can change the individual game in the MAME menu creating a specific config for said game.  But when you setup an arcade cabinet in my experience it works out best to set the default mappings for all games to street fighter layout.  In general that works for most games.  then if needed you change specific games.  Unfortunately the mamegenerator is automatically changing the default.cfg file in MAME and changing buttons for all in MAME will not work because mamegenerator

TVsIan commented:
In that case, I can just add in a “don’t overwrite config” option.

On Thu, Dec 16, 2021 at 1:58 PM ElMeroMaquinero ***@***.***>
wrote:

> You can, just not system-wide. Changing the options in the MAME menu under
> “this machine” will save them for that game. But I’ll see what FBNeo does,
> and maybe set up a default based on that if it’s different plus one for
> fightsticks.
> … <#m_-375410016967853981_>
>
> Yes you are correct. You can change the individual game in the MAME menu
> creating a specific config for said game. But when you setup an arcade
> cabinet in my experience it works out best to set the default mappings for
> all games to street fighter layout. In general that works for most games.
> then if needed you change specific games. Unfortunately the mamegenerator
> is automatically changing the default.cfg file in MAME and changing buttons
> for all in MAME will not work because mamegenerator
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996092697>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2R5JGLS7IBNSNKNGFTURIZGRANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ElMeroMaquinero commented:
So I managed to edit mamegenerator.py and got my street fighter buttons mapped correctly except for one... the coin2 mapping.  Looked all over mamegenerator.py and could not find where coin2 is mapped.

ElMeroMaquinero commented:
> In that case, I can just add in a “don’t overwrite config” option.
> […](#)

I was in the forums a few days ago asking the same thing.. why not just have the option to turn off

TVsIan commented:
The coin and start buttons aren’t mapped, the system applies keymaps to set
controller 1-4’s start and select buttons to keyboard keys 1-8. It was like
that before I made any changes, and I didn’t want to mess with it.

On Thu, Dec 16, 2021 at 2:05 PM ElMeroMaquinero ***@***.***>
wrote:

> So I managed to edit mamegenerator.py and got my street fighter buttons
> mapped correctly except for one... the coin2 mapping. Looked all over
> mamegenerator.py and could not find where coin2 is mapped.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996097422>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2VFBYM4L236IWV5JVDURI2AHANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ElMeroMaquinero commented:
The buttons are setup correctly in ES but for some reason the select button for player 2 is not translating to coin2 in MAME

TVsIan commented:
That would be an issue with the pad2key then. I would just wait until I
update this, then set your controls inside MAME.

On Thu, Dec 16, 2021 at 2:10 PM ElMeroMaquinero ***@***.***>
wrote:

> The buttons are setup correctly in ES but for some reason the select
> button for player 2 is not translating to coin2 in MAME
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996101080>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2REVSWUU5YR7WLVDZDURI2TBANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ElMeroMaquinero commented:
> That would be an issue with the pad2key then. I would just wait until I update this, then set your controls inside MAME.
> […](#)

So your update will have the option to disable autoconfig?

TVsIan commented:
Yeah, I’ll get that added in tonight.

On Thu, Dec 16, 2021 at 2:19 PM ElMeroMaquinero ***@***.***>
wrote:

> That would be an issue with the pad2key then. I would just wait until I
> update this, then set your controls inside MAME.
> … <#m_-4999798389905652435_>
>
> So your update will have the option to disable autoconfig?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996108293>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2TYBFN3W43XPETSEDTURI3UZANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ElMeroMaquinero commented:
> Yeah, I’ll get that added in tonight.
> […](#)

👍🏾

TVsIan commented:
Building and testing now. I set Neo Geo games to use the Neo Geo Mini pad layout (West = A, South = B, North = C, East = D), that seemed like the closest to modern controls, and makes KoF's face buttons match up with the Steam version of KoF XIII - punches on top, kicks on the bottom.

The Neo Geo CD pad is basically rotated 90 degrees - South = A, East = B, West = C, North = D. That looks like it only makes sense if you're used to that pad.

And Neo Geo Fightstick is based on a modern 8-button fightstick - South = A, North = B, R1 = C, R2 = D. That way, your fingers are on the top row of buttons and your thumb is one row lower, like the arcade cabinet. It wouldn't make sense for any game with > 4 buttons, but might be nice to use for ones with 3 or less on a stick.

The "Custom controls" option for MESS system has been renamed "Custom Game Controls," because the option "Custom MAME controls" will simply turn off overwriting any config that already exists. Custom Game Controls will still create a copy of the configs in a folder named after the game to allow per-game controls on MESS ROMs. Custom MAME controls would allow per-system custom controls on MESS systems as well, since it won't overwrite the MESS system config either.

The recommendation would be to use that option on arcade cabinets or other non-standard controllers that won't change, and leave it off when using standard controllers or changing controllers often.

Hew-ux commented:
Ah, https://github.com/batocera-linux/batocera.linux/issues/4889 strikes again.
You've had your work cut out for you! Remember to update the label if functionality isn't self-evident.

TVsIan commented:
> Ah, #4889 strikes again. You've had your work cut out for you! Remember to update the label if functionality isn't self-evident.

The only change I made to get everything working for the modified MAME generator was to pass only the direction that's defined to the one function that was using both (the CD-i directional control mapping). The generator was originally built to ignore that issue and just go by what the name of the direction says it should be.

I also just spun the lists of games to automap into a series of csv files that the generator should read in, which makes changing them or adding future automaps a bit cleaner.

Waiting on the image to finish building so I can test this out. Then finally do a last test on the mega bezels and push that out for testing.

TVsIan commented:
Okay, this should be ready to go, or at least stable and ready for testing. The complete list of changes (I think):

- Added multiple MAME controller presets intended for fighting games, to be set manually or applied automatically if set to Auto:
- Default: Current Batocera default
- Street Fighter (SNES): Default for Capcom 6-button games
- Street Fighter (Modern): Intended for fightsticks
- Mortal Kombat (SNES): Default for Mortal Kombat series & clones
- Killer Instinct (SNES): Default for Killer Instinct games
- Genesis 6-Button (Retroarch): Intended to match Genesis core mapping for 6-button controllers
- Neo Geo (Neo Geo Mini Pad): Default for Neo Geo games
- Neo Geo (Neo Geo CD Pad)
- Neo Geo (Offset Fightstick): Intended for fightsticks to match the arcade layout (A in bottom left, top right 3 buttons for B, C, D)
- Fixed analog triggers mapping to buttons (possibly XInput only, needs testing)
- Added "Custom MAME Configuration" option to not overwrite any existing default.cfg or MESS system cfg. Note that if turned off for ANY MAME/MESS system, it will overwrite the shared configs. This is an advanced setting intended for arcade cabinets and other non-standard controller layouts.
- All auto-preset game lists in text files for future changes if needed
- Config generation code split from main generator code
- Cleaned up MESS controller generation code, should allow for future alternate mappings if needed.

ElMeroMaquinero commented:
That’s kool 👍🏾

Please excuse my ignorance as I’m not familiar with how these pull requests get from here to batocera build.  Do I simply update to the latest beta?

TVsIan commented:
Once it gets merged, it should show up in the next beta after that.

If you really wanted to, you could download the source and add and compile
it yourself, but the easy way would be just to wait.

On Fri, Dec 17, 2021 at 12:35 AM ElMeroMaquinero ***@***.***>
wrote:

> That’s kool 👍🏾
>
> Please excuse my ignorance as I’m not familiar with how these pull
> requests get from here to batocera build. Do I simply update to the latest
> beta?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996451160>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2VBTQEXVNP5P66N4YLURLDY7ANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Hew-ux commented:
>Custom MAME Configuration [...] Note that if turned off for ANY MAME/MESS system, it will overwrite the shared configs.

I think this might need to be changed. In fact, @NostalgicDavid just encountered this situation too.

To quote myself: 
>So something like "if user has set 'use batocera keymaps' to false:
>read the current keymaps from X file...
>apply keymaps to retroarchOptions.cfg...
>continue to launch game
>else:
>apply preset"

The full conversation starts at https://discord.com/channels/357518249883205632/524321176068161554/920817090993193011

Iconocluster's suggestion for storing the custom user mapping in batocera.conf is a good idea.

The simplest solution for this PR is to just not include a "Custom MAME configuration" option yet, and to implement that in another PR.

TVsIan commented:
Actually, one thing I just thought of is that I can have each system used
it’s own config path (just a sub folder in the MAME config folder), so that
will solve that particular issue.

I’ll make that change after work today, someone else can expand on that
because I think that other idea is beyond me.

On Fri, Dec 17, 2021 at 6:36 AM Atari ***@***.***> wrote:

> Custom MAME Configuration [...] Note that if turned off for ANY MAME/MESS
> system, it will overwrite the shared configs.
>
> I think this might need to be changed. In fact, @NostalgicDavid
> <https://github.com/NostalgicDavid> just encountered this situation too.
>
> To quote myself:
>
> So something like "if user has set 'use batocera keymaps' to false:
> read the current keymaps from X file...
> apply keymaps to retroarchOptions.cfg...
> continue to launch game
> else:
> apply preset"
>
> The full conversation starts at
> https://discord.com/channels/357518249883205632/524321176068161554/920817090993193011
>
> Iconocluster's suggestion for storing the custom user mapping in
> batocera.conf is a good idea.
>
> The simplest solution for this PR is to just not include a "Custom MAME
> configuration" option yet, and to implement that in another PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-996654999>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2QSO5X5UL5HD6HOAZ3URMOCRANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TVsIan commented:
Okay, made a few more changes and testing after work.

- Changed the generator to use a subdirectory for each MAME system, so turning custom config on/off per system won't affect any of the others. Turning it on for a single game would overwrite it for that system, but there's absolutely no reason to do that since MAME will store a per-game config and there's an option to do that for the MESS systems.
- Renamed the control option from Fighting Game Layout to Special Control Layouts, because I also added:
- Twinstick with Button controls (Based on the currently not emulated right Virtual On) - 1 = L2, 2 = L1, 3 = R2, 4 = R1, 5 = L3, 6 = R3 - for games that would have had two sticks with buttons or triggers on the sticks themselves. For VON, buttons 1 & 3 are the triggers and 2 & 4 are the buttons, so they're mapped to the closest physical button. 5 & 6 are unused there, but I saw at least one 6 button game (the unreleased Air Race prototype), so I put them on the stick clicks.
- Rotated 4-Way Stick (Q*Bert series, Congo Bongo): This will map the directional controls to the diagonals that match the direction the character moves. For example, MAME's Up in Q*Bert will move up and right, so the Up direction is mapped to Up & Right together. Works for both analog & D-Pad.

Both of those will map automatically to the handful of games that would want them.

Those are the only other control types that wouldn't map properly that I can think of. Maybe mapping a gear shift to the right stick for some driving games, but unless there's demand for it, I'll stop here. If it looks like everything is Ok, I'll push another commit.

nadenislamarre commented:
thanks a lot for this big work.

files limited to 1 emulator should be defined in the generator, not in this file (we target to clean this file)
package/batocera/core/batocera-configgen/configgen/configgen/batoceraFiles.py

this is not a problem for this pr, just an info.

TVsIan commented:
I have one final set of changes that I'll submit in a separate PR, I'll move them from that file when I do.

Hew-ux commented:
Wow! You really know how to outdo yourself. Wonderful work.

I don't like how the user can still accidentally rekt themself with the custom mame config option but at least it is labelled somewhat appropriately. But as you said, WIP so I will wait and see what you have in store for us.

palmerj commented:
@TVsIan Thanks, heaps for these changes, much needed with all of the fighting games.

I've been using this for a week and loving it! I mostly use the Genesis 6-Button (Retroarch) layout to match my 8BitDo M30 controller. The only issue I'm having is I play Mortal Kombat, Street Fighter and Killer Instinct all the time. The layout you have implemented, "Genesis 6-Button", worked correctly for Street Fighter/Capcom and Killer instinct games, but not for Mortal Kombat. For Mortal Kombat the run and low kick buttons are swapped. So this means another alternative layout where buttons 5 and 6 are swapped is needed. E.g:

```
# Genesis-style controller layout (Based on Mortal Kombat 3)
mappings.update({"BUTTON1": "pageup"})
mappings.update({"BUTTON2": "x"})
mappings.update({"BUTTON3": "pagedown"})
mappings.update({"BUTTON4": "y"})
mappings.update({"BUTTON5": "a"})
mappings.update({"BUTTON6": "b"})
```

However, given this feature is not officially released yet, I wonder about a better way to implement this alternative configuration system, given you already have 3 SNES layouts for MK, SF, and KI. Maybe another approach is to define a layout (e.g Neo Geo, SNES, Genesis) and then automatically set the button mappings based on the game, which you already have detection code. This would then be more extensible for other layouts and reduce the number of entries in the user list. I'm not sure about the impact of the Twinstick layout you have done, but I assume that would be ok in this case, as you wouldn't change the mapping unless it matches a Twinstick game. I can submit a PR to add this layout but need some advice here. 

Also sorry to open the conversation on this again!

Jeremy

TVsIan commented:
Yeah, that might be the way to go. I think the final menu would probably be
something like SNES, Genesis, Fightstick, then the two Neo Geo pad options,
Twinstick, and Q*Bert.

I can rewrite it when I have a minute (working on some other changes right
now), but generally I would think:
Auto or SNES - current auto behavior
Genesis - Current Genesis for Capcom & KI, new layout for MK, ?? layout for
other (likely a flipped SFII would make sense, 1-3 on A-C, 4-6 on X-Z)
Fightstick - Current Modern SFII on Capcom, ?? for KI and MK (need to look
up what the current versions do on a stick), current Neo Geo fightstick,
new layout for other (1-4 bottom row, 5-8 top row)
Neo Geo Mini Pad - Unchanged
Neo Geo CD Pad - Unchanged
Twinstick - Unchanged
Q*Bert - Unchanged

The new layouts would have internal numbers like the current ones, but
rather than set them manually in the options, the generator would flip it
when IDing the game. I'll move the game detection to a new function to
clean things up a bit.

On Thu, Jan 6, 2022 at 1:30 AM Jeremy Palmer ***@***.***>
wrote:

> @TVsIan <https://github.com/TVsIan> Thanks, heaps for these changes, much
> needed with all of the fighting games.
>
> I've been using this for a week and loving it! I mostly use the Genesis
> 6-Button (Retroarch) layout to match my 8BitDo M30 controller. The only
> issue I'm having is I play Mortal Kombat, Street Fighter and Killer
> Instinct all the time. The layout you have implemented, "Genesis 6-Button",
> worked correctly for Street Fighter/Capcom and Killer instinct games, but
> not for Mortal Kombat. For Mortal Kombat the run and low kick buttons are
> swapped. So this means another alternative layout where buttons 5 and 6 are
> swapped is needed. E.g:
>
> # Genesis-style controller layout (Based on Mortal Kombat 3)
> mappings.update({"BUTTON1": "pageup"})
> mappings.update({"BUTTON2": "x"})
> mappings.update({"BUTTON3": "pagedown"})
> mappings.update({"BUTTON4": "y"})
> mappings.update({"BUTTON5": "a"})
> mappings.update({"BUTTON6": "b"})
>
> However, given this feature is not officially released yet, I wonder about
> a better way to implement this alternative configuration system, given you
> already have 3 SNES layouts for MK, SF, and KI. Maybe another approach is
> to define a layout (e.g Neo Geo, SNES, Genesis) and then automatically set
> the button mappings based on the game, which you already have detection
> code. This would then be more extensible for other layouts and reduce the
> number of entries in the user list. I'm not sure about the impact of the
> Twinstick layout you have done, but I assume that would be ok in this case,
> as you wouldn't change the mapping unless it matches a Twinstick game. I
> can submit a PR to add this layout but need some advice here.
>
> Also sorry to open the conversation on this again!
>
> Jeremy
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-1006318535>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2V5O3WNHXZM7KVMYBTUUUZI3ANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

palmerj commented:
Hi @TVsIan, thanks for the reply. 

> Yeah, that might be the way to go. I think the final menu would probably be
something like SNES, Genesis, Fightstick, then the two Neo Geo pad options,
Twinstick, and Q*Bert.

Great!

>  layout for other (likely a flipped SFII would make sense, 1-3 on A-C, 4-6 on X-Z)

What would be the purpose of this?

> The new layouts would have internal numbers like the current ones, but
rather than set them manually in the options, the generator would flip it
when IDing the game. 

Sure. I was also thinking that the values for altlayout in the es_features.cfg shouldn't be numbers, but actual string values. e.g:

```
    <feature name="FIGHTING GAME LAYOUT" value="altlayout" description="Controls for 5/6 button games">
      <choice name="SNES" value="snes" />
      <choice name="fightstick" value="fightstick" />
      <choice name="Genesis" value="genesis" />
      <choice name="Neo Geo CD" value="neogeo_cd" />
      <choice name="Neo Geo Fightstick" value="neogeo_fightstick" />
      <choice name="Twinstick with Buttons (Battle Zone, virtual On)" value="twinstick" />
    </feature>
```

This way it makes the parameter management easier to deal with in code and within batocera.conf. 

> I'll move the game detection to a new function to
clean things up a bit.

Sounds good.

TVsIan commented:
> Hi @TVsIan, thanks for the reply.
> 
> > Yeah, that might be the way to go. I think the final menu would probably be
> > something like SNES, Genesis, Fightstick, then the two Neo Geo pad options,
> > Twinstick, and Q*Bert.
> 
> Great!
> 
> > layout for other (likely a flipped SFII would make sense, 1-3 on A-C, 4-6 on X-Z)
> 
> What would be the purpose of this?

Typically on an arcade control panel, buttons 1-4 will be the bottom row of buttons. For whatever reason, Capcom flipped it for their setup. So for most arcade games, if you're on a fightstick layout, you'll want the action buttons on the bottom row.

> > The new layouts would have internal numbers like the current ones, but
> > rather than set them manually in the options, the generator would flip it
> > when IDing the game.
> 
> Sure. I was also thinking that the values for altlayout in the es_features.cfg shouldn't be numbers, but actual string values. e.g:
> 
> ```
>     <feature name="FIGHTING GAME LAYOUT" value="altlayout" description="Controls for 5/6 button games">
>       <choice name="SNES" value="snes" />
>       <choice name="fightstick" value="fightstick" />
>       <choice name="Genesis" value="genesis" />
>       <choice name="Neo Geo CD" value="neogeo_cd" />
>       <choice name="Neo Geo Fightstick" value="neogeo_fightstick" />
>       <choice name="Twinstick with Buttons (Battle Zone, virtual On)" value="twinstick" />
>     </feature>
> ```
> 
> This way it makes the parameter management easier to deal with in code and within batocera.conf.

It would also break any configs that are already set to use it, and while it's still in beta I'd prefer not to do that.

> > I'll move the game detection to a new function to
> > clean things up a bit.
> 
> Sounds good.

palmerj commented:
> Typically on an arcade control panel, buttons 1-4 will be the bottom row of buttons. For whatever reason, Capcom flipped it for their setup. So for most arcade games, if you're on a fightstick layout, you'll want the action buttons on the bottom row.

Oh so this is for a fightstick, not a genesis layout? Sorry I got confused. 

> It would also break any configs that are already set to use it, and while it's still in beta I'd prefer not to do that.

Hmm. Just seems bad practise to use numbers here, and I would say only a handful of people will have used this so far. Given the other changes proposed here, I think what already done would be broken anyway right?

Hew-ux commented:
It's completely fine to rename option values if it makes it clearer. Worst thing that happens is that the user has to set the option again: not a biggie considering this hasn't even gone to stable yet.

TVsIan commented:
Ok then, once I knock out the glcore and Daphne changes, I'll loop back to
that. I'm also trying to add minivmac, but I suspect the libretro port may
not be good enough to include.

On Thu, Jan 6, 2022 at 4:20 PM Atari ***@***.***> wrote:

> It's completely fine to rename option values if it makes it clearer. Worst
> thing that happens is that the user has to set the option again: not a
> biggie considering this hasn't even gone to stable yet.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/5186#issuecomment-1006943058>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2RBXJZTRM6US7AITT3UUYBSRANCNFSM5JKFRBFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

palmerj commented:
Awesome thank so much @TVsIan :)

palmerj commented:
Can you please also support a Mortal Kombat mapping for the fightstick layout as well. I have a Hori Fighting Commander One which has the same fightstick layout, but also needs the 5th and 6th buttons swapped. Here's my mapping for batocera e.g

```
A = B
B = A
X = Y
Y = X
L1 = LB
R2 = RB
LT = LT
RT = RT
```
I think you will find that this mapping will be the same for all Xbox based fightsticks/fightpads. It think will also be the same for Playstation fightsticks because there are usually:

```
A = Circle
B = X
X = Triangle
Y = Square
L1 = L1
R2 = R2
LT = L2
RT = L2
```

I'm wondering if there is some mapping to do with neogeo too, but haven't tested this as I don't like the games too much. But in my mind it would be good to map the 4 buttons the a, b, x and y buttons.

palmerj commented:
I also had a friend come over and we started playing 2 player fighter games with 2 different pads. Can we please support mappings for 2 players with 2 different pad layouts?

TVsIan commented:
> I also had a friend come over and we started playing 2 player fighter games with 2 different pads. Can we please support mappings for 2 players with 2 different pad layouts?

Honestly, at that point you’re better off using the built in MAME options. It would be quite a bit of work to make that a per-player option for something that’s really an edge case.

TVsIan commented:
Got a version ready to test. If everything works, the final menu selection will be:

- Auto/unselected (will use the SNES mappings by default, since most controllers will be 4 face buttons + shoulders)
- Default (Batocera's default controls only, no changes)
- SNES Style
- Genesis/Megadrive Style
- Fightstick
- Neo Geo Mini Pad (default for Neo Geo games if on auto)
- Neo Geo CD Pad
- Twinstick
- Q*Bert

I did modify the MK Genesis style from the one posted here - that one had run & block flipped based on what the manual was showing. For MK Fightstick, there's not really an official version for a standard 8-button stick, so I kind of hybridized the arcade panel and Genesis layout - bottom row is run, lp, block, lk, top row is unused, hp, unused, hk. So block is on the bottom between the low attacks like Genesis, and run is off to the left for your thumb like the arcade.

I should have a pr ready in the next day or two, once I've had a chance to test things out.

palmerj commented:
Many thanks @TVsIan :)

> I did modify the MK Genesis style from the one posted here - that one had run & block flipped based on what the manual was showing

Can you please revert this back to what I posted (block top row, run bottom row)? Most 6 face button people will expect this, especially with a genesis style pad. This dates back to the implementation of MK3 on the genesis itself. See https://segaretro.org/index.php?title=File:Ultimate_Mortal_Kombat_3_MD_US_Manual.pdf&page=9. Also, you can see with pros it's also a pretty common layout https://www.youtube.com/watch?v=alZngCMQg7M&ab_channel=MgoUmk. It makes natural sense because the run button is lower than the block button on the arcade panel.

> For MK Fightstick, there's not really an official version for a standard 8-button stick, so I kind of hybridized the arcade panel and Genesis layout - bottom row is run, lp, block, lk, top row is unused, hp, unused, hk. So block is on the bottom between the low attacks like Genesis, and run is off to the left for your thumb like the arcade.

I guess this is ok for 8 button layouts, as that kinda tries to simulate the run button to the lower left of the low punch button as ob the official panel. However, I would suggest this is still a little weird when playing MK1 and MK2, which don't have a run button. I've mostly seen the top row mapping the middle button to block to allow more natural access to block on both the top and bottom rows. See

![image](https://user-images.githubusercontent.com/580022/148657549-bf70e85d-03b6-4330-bc67-b34fb5e3258c.png)

Lastly, I have fight stick and fight pad that only have 6 face buttons. e.g 

![image](https://user-images.githubusercontent.com/580022/148658133-76c84c99-6be9-412a-b3b0-e40aa6638bfb.png)

This is a standard layout for xinput fighting controllers. With the proposed MK Fightstick option of shifting the button across and making the 'X' face button unused and the face 'A' button run, it's not usable. Can we make at least have 6 button Fightstick option? In saying this, I really do think that just simplifying to top row is hp, block, hk, unused, and the bottom row is lp, run/block, lk, unused and having a single option for 6 and 8 button fightsticks is easier though :)

palmerj commented:
> Honestly, at that point you’re better off using the built in MAME options. It would be quite a bit of work to make that a per-player option for something that’s really an edge case.

Ok. I'll see how I go in the future. If the itch takes me after you have submitted this improved PR, then I might look at implementing for 2 players. 

Again many thanks for all of this, it will be a massive improvement for fighting gamers!

TVsIan commented:
Ok, flipped the buttons for Genesis, and basically made the fightstick version match.
